firmware/src/opaygo_functions.hpp

```diff
@@ -0,0 +1,313 @@
+                                               // proper nb of disconnections
+  nbDisconnections++;
+  storeNbDisconnectionsEEPROM();
+}
+
+void ChangeLedState(int ledPin) {
+  digitalRead(ledPin) == LOW ? digitalWrite(ledPin, HIGH)
+                             : digitalWrite(ledPin, LOW);
+}
+
+void BlinkLED(int LedPin, int NumberOfBlinks, int BlinkPeriode) {
+  int i;
+  for (i = 0; i < NumberOfBlinks; i++) {
+    ChangeLedState(LedPin);
+    delay(BlinkPeriode);
+    ChangeLedState(LedPin);
+    delay(BlinkPeriode);
+  }
+}
+
+void BlinkRedLED(int NumberOfBlinks, int BlinkPeriode) {
+  BlinkLED(red_led, NumberOfBlinks, BlinkPeriode);
+}
+
+void BlinkGreenLED(int NumberOfBlinks, int BlinkPeriode) {
+  BlinkLED(green_led, NumberOfBlinks, BlinkPeriode);
+}
+
+uint32_t GetTimeInSeconds() {
+  /*
+   *Returns the unixtime in seconds of Now
+   */
+  DateTime now = rtc.now();
+  uint32_t nowInSeconds =
+      (now.unixtime() -
+       timeInitializationRtc);  // we substract the init time so that it's
+                                // easier to read when debugging, and works the
+                                // same as Arduino Time mgt
+  return (nowInSeconds);
+}
+
+bool TokenEntryAllowed() {
+  if (TokenEntryLockedUntil > GetTimeInSeconds()) {
+    return false;
+  } else {
+    return true;
+  }
+}
+
+int GetKeyPressed()
+/*
+ * returns the key pressed on the Keypad remote as a int (this function forces
+ * to put the key in a char, good practice)
+ */
+{
+  char incomingByte;
+  incomingByte = customKeypad.getKey();
+  Serial.print("incoming byte = ");
+  Serial.println(incomingByte);
+  switch (incomingByte) {
+    case '*':
+      return STAR_KEY;
+      break;
+
+    case '#':
+      return HASH_KEY;
+      break;
+
+    default:
+      return (int)(incomingByte - '0');  // this_char is now an int
+      break;
+  }
+}
+
+void UpdateInvalidTokenWaitingPeriod() {
+  uint32_t Now = GetTimeInSeconds();
+
+  // We check that it does not become unbearably long
+  if (InvalidTokenCount > 11) {
+    InvalidTokenCount = 11;
+  }
+
+  // We add some forgiveness for the first 2 errors
+  if (InvalidTokenCount > 2) {
+    TokenEntryLockedUntil = Now + pow(2, InvalidTokenCount - 2) * 60;
+  }
+}
+
+void AddTime(int ActivationValue) {
+  uint32_t Now = GetTimeInSeconds();
+  int NumberOfSecondsToActivate = (ActivationValue * 3600 * 24) / TIME_DIVIDER;
+
+  if (ActiveUntil < Now) {
+    ActiveUntil = Now;
+  }
+
+  ActiveUntil +=
+      NumberOfSecondsToActivate;  // We add the number of days (converted in
+                                  // seconds for to compare to our RTC time)
+}
+
+void SetTime(int ActivationValue) {
+  uint32_t Now = GetTimeInSeconds();
+  int NumberOfSecondsToActivate = (ActivationValue * 3600 * 24) / TIME_DIVIDER;
+
+  ActiveUntil =
+      Now +
+      NumberOfSecondsToActivate;  // We set the number of days (converted in
+                                  // seconds for to compare to our RTC time)
+}
```

Comment:
This isn't PAYGO related, but time related, should it go in `time_management.hpp` instead?